core/src/main/java/com/linecorp/armeria/server/ServerHttp2ObjectEncoder.java

```diff
@@ -82,22 +83,32 @@ public boolean isResponseHeadersSent(int id, int streamId) {
     }
 
     private Http2Headers convertHeaders(HttpHeaders inputHeaders, boolean isTrailersEmpty) {
-        final Http2Headers outHeaders = ArmeriaHttpUtil.toNettyHttp2ServerHeaders(inputHeaders);
-        if (!isTrailersEmpty && outHeaders.contains(HttpHeaderNames.CONTENT_LENGTH)) {
-            // We don't apply chunked encoding when the content-length header is set, which would
-            // prevent the trailers from being sent so we go ahead and remove content-length to force
-            // chunked encoding.
-            outHeaders.remove(HttpHeaderNames.CONTENT_LENGTH);
+        HttpHeadersBuilder builder = null;
+        if (enableServerHeader && !inputHeaders.contains(HttpHeaderNames.SERVER)) {
+            builder = inputHeaders.toBuilder();
+            builder.add(HttpHeaderNames.SERVER, ArmeriaHttpUtil.SERVER_HEADER);
         }
 
-        if (enableServerHeader && !outHeaders.contains(HttpHeaderNames.SERVER)) {
-            outHeaders.add(HttpHeaderNames.SERVER, ArmeriaHttpUtil.SERVER_HEADER);
+        if (enableDateHeader && !inputHeaders.contains(HttpHeaderNames.DATE)) {
+            if (builder == null) {
+                builder = inputHeaders.toBuilder();
+            }
+            builder.add(HttpHeaderNames.DATE, HttpTimestampSupplier.currentTime());
         }
 
-        if (enableDateHeader && !outHeaders.contains(HttpHeaderNames.DATE)) {
-            outHeaders.add(HttpHeaderNames.DATE, HttpTimestampSupplier.currentTime());
+        if (!isTrailersEmpty && inputHeaders.contains(HttpHeaderNames.CONTENT_LENGTH)) {
+            if (builder == null) {
+                builder = inputHeaders.toBuilder();
+            }
+            // We don't apply chunked encoding when the content-length header is set, which would
+            // prevent the trailers from being sent so we go ahead and remove content-length to force
+            // chunked encoding.
+            builder.remove(HttpHeaderNames.CONTENT_LENGTH);
+        }
+        if (builder == null) {
+            builder = inputHeaders.toBuilder();
```

Comment:
How about assigning this builder at the first of the method and remove all `if (builder == null)`?

Reply:
Oops... I don't remember why I did like this.🤣
Maybe the code might be changed while fixing test cases.